### PR TITLE
Parse try, comma and pipe with correct precedence

### PR DESCRIPTION
**Pull request by wader**

### CODE REVIEW

**Comment by wader** [comment]
@thaliaarchi looks correct?

File: jqjq.test [hunk 1]
```diff
@@ -659,6 +659,22 @@ try error(.) catch .
...
+try 1 catch error, try error catch 3
+null
+1
+3
+
+try 1 catch 2 | 3
+null
+3
+
+# TODO: some try 1 | ... test?
```
Comment (wader) [inline jqjq.test:676]: didn't come up with a good test

**Comment by thaliaarchi** [comment]
The code and tests look good, but I since I'm not familiar with this part of the grammar, I looked at a few implementations to `try` and understand the semantics.

Looking at jq's parser doesn't make it clear to me if this is correct, though. jq defines `try`/`catch` [parser.y](https://github.com/jqlang/jq/blob/c95b34ff827d05a2d262f00280a4891a295ed0ed/src/parser.y#L359-L368) as such:

```yacc
"try" Exp "catch" Exp {
  $$ = gen_try($2, $4);
} |
"try" Exp {
  $$ = gen_try($2, gen_op_simple(BACKTRACK));
} |
"try" Exp "catch" error {
  FAIL(@$, "Possibly unterminated 'try' statement");
  $$ = $2;
} |
```

It simply uses `Exp`, so it doesn't clearly (to me) delineate which tokens are forbidden in that context. I don't think yacc considers the order of definitions, as it's bottom-up. I'm not sure how to learn the precise definition used in practice.

[gojq](https://github.com/itchyny/gojq/blob/e47f1f985947ba2a8ff1c6cdcf8b4a34f4e96cf1/parser.go.y#L387-L390) also uses a parser generator and has the same issues. [jaq](https://github.com/01mf02/jaq/blob/18b4bad1b35be1fcf99175fc75c23d17d654e742/jaq-parse/src/filter.rs#L49-L57) uses parser combinators (with Chumsky) and it does not seem to specially consider what punctuation is allowed in `try`/`catch` with [`ignore_then`](https://docs.rs/chumsky/latest/chumsky/trait.Parser.html#method.ignore_then).

How did you arrive at your definition? Is it common knowledge, that isn't in code anywhere?

Ultimately, if it improves test coverage with jq's tests (which it does), then I'd say it's good.

**Comment by wader** [comment]
> The code and tests look good, but I since I'm not familiar with this part of the grammar, I looked at a few implementations to `try` and understand the semantics.
> 
> Looking at jq's parser doesn't make it clear to me if this is correct, though. jq defines `try`/`catch` [parser.y](https://github.com/jqlang/jq/blob/c95b34ff827d05a2d262f00280a4891a295ed0ed/src/parser.y#L359-L368) as such:
> 
> ```yacc
> "try" Exp "catch" Exp {
>   $$ = gen_try($2, $4);
> } |
> "try" Exp {
>   $$ = gen_try($2, gen_op_simple(BACKTRACK));
> } |
> "try" Exp "catch" error {
>   FAIL(@$, "Possibly unterminated 'try' statement");
>   $$ = $2;
> } |
> ```
> 
> It simply uses `Exp`, so it doesn't clearly (to me) delineate which tokens are forbidden in that context. I don't think yacc considers the order of definitions, as it's bottom-up. I'm not sure how to learn the precise definition used in practice.
> 
> [gojq](https://github.com/itchyny/gojq/blob/e47f1f985947ba2a8ff1c6cdcf8b4a34f4e96cf1/parser.go.y#L387-L390) also uses a parser generator and has the same issues. [jaq](https://github.com/01mf02/jaq/blob/18b4bad1b35be1fcf99175fc75c23d17d654e742/jaq-parse/src/filter.rs#L49-L57) uses parser combinators (with Chumsky) and it does not seem to specially consider what punctuation is allowed in `try`/`catch` with [`ignore_then`](https://docs.rs/chumsky/latest/chumsky/trait.Parser.html#method.ignore_then).
> 
> How did you arrive at your definition? Is it common knowledge, that isn't in code anywhere?

Thanks! that is way more research then i did :) will study them closer tomorrow or so. I happened to noticed that jqjq behaved differently while writing some of the invalid path tests when doing `try ..., try...`. Don't think there is any common knowledge really and now i see that it's probably all infix operators that are not allowed? ex `try 1 + 2 catch 3` does not parse with jq, but feel like it would have been nice if it did work? maybe causes other issues?

I'm also a bit confusing how yacc works, had a quick look at jq's and gojq's parsers i'm not sure i follow why it does not works to parse `try 1 + 2 catch 3`, feels like it cab be valid parse tree? i must be missing something

> Ultimately, if it improves test coverage with jq's tests (which it does), then I'd say it's good.

Yeap, think i will improve it a bit more then merge.

btw if you want to see how gojq parses something you can do this. Also jqjq uses nearly the exact same AST (some small differences) which makes it quite easy to compare. I wonder if it's possible to get some kind of parse tree out of jq? yacc in debug mode somehow?

``` sh
$ fq -n '"try 1 catch 2 | try 3 catch 4" | _query_fromstring'
{
  "left": {
    "term": {
      "try": {
        "body": {
          "term": {
            "number": "1",
            "type": "TermTypeNumber"
          }
        },
        "catch": {
          "term": {
            "number": "2",
            "type": "TermTypeNumber"
          }
        }
      },
      "type": "TermTypeTry"
    }
  },
  "op": "|",
  "right": {
    "term": {
      "try": {
        "body": {
          "term": {
            "number": "3",
            "type": "TermTypeNumber"
          }
        },
        "catch": {
          "term": {
            "number": "4",
            "type": "TermTypeNumber"
          }
        }
      },
      "type": "TermTypeTry"
    }
  }
}
```

**Comment by wader** [comment]
Had forgot that "query1" parser name (horrible name) matches things that are not infix operators so i guess that is what we want. But i do wonder how to do tests that don't have to be based on parsing and assert the ast for some of these, ex is there a way to make  `try X + Y` behave differently if parsed as `(try X) + Y` vs `try (X + Y)` 🤔

**Comment by thaliaarchi** [comment]
AST tests do sound nice. Maybe you could try jqunit, since those seem more like unit tests than integration tests that would be reusable with other implementations.

**Comment by wader** [comment]
Interesting idea, maybe reimplement something jqunit-ish 🤔 

```sh
$ cat test.jq
def should_pass_test: 1;

def should_fail_test: error("Fails test");

def nope: 123;

$ cat test.jq | jq -cRs -L . 'include "jqjq"; lex | parse | .func_defs[] | select(.name | startswith("should_")) | .name, try ["ok", eval_ast(.body)[1]] catch ["error", .]'
"should_pass_test"
["ok",1]
"should_fail_test"
["error","Fails test"]
```

I've thought about implement AST to string support and if that could be configured to add explicit parentheses it could maybe be used to build convenient AST tests.

**Comment by thaliaarchi** [comment]
Heh, somehow I shouldn’t be surprised that you want to reimplement this in jq, but I was blindsided on this one and didn’t expect it, because I like Rust and usually rewrite code in the other direction 😂.

I took a look at jqunit to see if they’re multi-threaded. It’s not, but it seems like it would be simple to `jq_init` once in each thread. Multi-threading would make running your tests much faster.

It would also be fantastic if jqunit could run tests in many engines (jq, gojq, jaq, jqjq) and report differences.

But jqunit is currently tiny and would be easy to reimplement in jq.

**Comment by wader** [comment]
> Heh, somehow I shouldn’t be surprised that you want to reimplement this in jq, but I was blindsided on this one and didn’t expect it, because I like Rust and usually rewrite code in the other direction 😂.

hehe when it comes jqjq everything looks like a jq problem to me 😬

> I took a look at jqunit to see if they’re multi-threaded. It’s not, but it seems like it would be simple to `jq_init` once in each thread. Multi-threading would make running your tests much faster.
> 
> It would also be fantastic if jqunit could run tests in many engines (jq, gojq, jaq, jqjq) and report differences.

Yeap that would be nice. There was some talks about some kind of common jq test suits in https://github.com/01mf02/jq-lang-spec/issues/1 but i do also like the current .test-format a lot, or at least the idea to have a convenient standardised format that jq implementations can implement.

> But jqunit is currently tiny and would be easy to reimplement in jq.

Yeap seems so!

**Pull request merged**